File: LinkedList/LinkedList.py
# Lista bidirezionale circolare con sentinella
class LinkedList:
    class Node:
        def __init__(self, value):
            self.value = value
            self.pred = self
            self.succ = self

    def __init__(self):
        self.sentinel = self.Node(None)

    def head(self):
        return self.sentinel.succ

    def tail(self):
        return self.sentinel.pred

    def next(self, p):
        return p.succ

    def prev(self, p):
        return p.pred

    def isEmpty(self):
        return self.sentinel.succ == self.sentinel.pred == self.sentinel

    def finished(self, p):
        return self.sentinel == p

    def append(self, value):
        t = self.Node(value)
        t.pred = self.tail()
        t.succ = self.sentinel
        self.tail().succ = t
        self.sentinel.pred = t
        return t

    def insert(self, p, value):
        t = self.Node(value)
        t.pred = p.pred
        p.pred.succ = t
        t.succ = p
        p.pred = t
        return t

    def remove(self, p):
        p.pred.succ = p.succ
        p.succ.pred = p.pred
        t = p.succ
        del p
        return t

    def read(self, p):
        return p.value

    def write(self, p, value):
        p.value = value

    def printAll(self, p):
        if p.value == None:
            return
        else:
            print(p.value)
            self.printAll(p.succ)


list = LinkedList()
# x1 = list.insert(list.head(), 1)
# x2 = list.insert(x1, 2)
# x3 = list.insert(x2, 3)
list.append(1)
list.append(2)
list.append(3)
print(f"TESTA: {list.head().value}")
print(f"CODA: {list.tail().value}")
list.printAll(list.head())

File: Containers/LinkedList.java
package Containers;

/**
 * List bidirectional, circular, with sentinel
 */

public class LinkedList<T> {
    private Node<T> sentinel;

    LinkedList() {
        this.sentinel = new Node<>(null);
    }

    public Node<T> head() {
        return this.sentinel.succ;
    }

    public Node<T> tail() {
        return this.sentinel.pred;
    }

    public Node<T> next(Node<T> pos) {
        return pos.succ;
    }

    public Node<T> prev(Node<T> pos) {
        return pos.pred;
    }

    public boolean isEmpty() {
        return (this.sentinel.succ == this.sentinel.pred) && (this.sentinel.succ == this.sentinel);
    }

    public boolean finished(Node<T> pos) {
        return this.sentinel == pos;
    }

    public Node<T> append(T value) {
        Node<T> t = new Node<>(value);
        t.pred = this.tail();
        t.succ = this.sentinel;
        this.tail().succ = t;
        this.sentinel.pred = t;
        return t;
    }

    public Node<T> insert(Node<T> pos, T value) {
        Node<T> t = new Node<>(value);
        t.pred = pos.pred;
        pos.pred.succ = t;
        t.succ = pos;
        pos.pred = t;
        return t;
    }

    public Node<T> remove(Node<T> pos) {
        pos.pred.succ = pos.succ;
        pos.succ.pred = pos.pred;
        Node<T> t = pos.succ;
        pos = null;
        return t;
    }

    public T read(Node<T> pos) {
        return pos.value;
    }

    public void write(Node<T> pos, T value) {
        pos.value = value;
    }

    public void printAll(Node<T> pos) {
        if (pos.value == null) {
            return;
        } else {
            System.out.print(pos.value + " ");
            this.printAll(pos.succ);
        }
    }
}
